### ISSUE TITLE
New OutputFilter: <script type="module">

### ISSUE THREAD

**Issue by lkraav** [inpsyde/assets#18]
**Is your feature request related to a problem? Please describe.**
ES Modules are a growing piece of the ecosystem. https://core.trac.wordpress.org/ticket/12009 still doesn't have a core solution.

**Describe the solution you'd like**
Add new OutputFilter here.

**Describe alternatives you've considered**
N/A

**Additional context**
N/A

**Comment by Chrico** [comment]
Thanks a lot for your input. I need a bit time to check this. I'm not so sure if an `OutputFilter` is the correct place to implement this. For me this is more a kind of "variation", so it should be configurable for example on the `Script` itself like `Script::withType(Script::TYPE_MODULE)` or maybe we can also have a new `ModuleScript`.

**Comment by lkraav** [comment]
> For me this is more a kind of "variation", so it should be configurable for example on the Script itself like Script::withType(Script::TYPE_MODULE) or maybe we can also have a new ModuleScript.

I don't know. I'm eyeballing this package's structure, both `ModuleScript` and new `Script::withType()` would be somewhat asymmetric to how existing system works.

I've also seen `add_script_add_data()` used like:

```php
global $wp_scripts;

wp_script_add_data( $handle, 'type', 'module' );

add_filter( 'script_loader_tag', static function( string $tag, string $handle ): string {
  if ( ! empty( $wp_scripts->registered[ $handle ]->extra['type'] ) ) {
    $tag = str_replace( '<script ', '<script type="' . esc_attr( $wp_scripts->registered[ $handle ]->extra['type'] ) . '" ', $tag );
  }
}
...

But at the end of the day, you're still with `str_replace`, so seems aligned with OutputFilter system here :thinking:

**Comment by nlemoine** [comment]
There's also some issues with the basic `str_replace` solution (#7)

Note that WordPress can also output a `type="text/javascript"` (see https://github.com/inpsyde/assets/pull/7#issuecomment-682719097). I don't how a browser would behave with both `type="text/javascript"` and `type="text/module"` 🤔

**Comment by lkraav** [comment]
> I think hardening the OutputFilter logic would be a place to start win

I actually have `preg_match()` based logic to account for existing `type` attribute, but cut it out because I wasn't sure how to build a test for it. Thanks for reminding me on how `type="text/javascript"` can be output, this may be enough motivation to refactor it back to work with both cases.

Hardening OutputFilter logic could boil down to moving to proper `DOMDocument`. Are you OK with bringing this dependency in?

**Comment by nlemoine** [comment]
> Hardening OutputFilter logic could boil down to moving to proper `DOMDocument`

That would the best way but it would not be as performant as a basic `str_replace` and it seems a bit overkill to me for such a task.

I started working a some ways to improve this that doesn't involve DOM parsing a while ago, I didn't have time to finish this properly. Adding the URL and/or handle (every asset has to have a unique handle -> `id='my-script'`) in the replacement, although not bullet proof, would harden things enough IMHO.

https://github.com/inpsyde/assets/compare/master...nlemoine:fix-output-filters

**Comment by saas786** [comment]
> > Hardening OutputFilter logic could boil down to moving to proper `DOMDocument`
> 
> That would the best way but it would not be as performant as a basic `str_replace` and it seems a bit overkill to me for such a task.
> 
> I started working a some ways to improve this that doesn't involve DOM parsing a while ago, I didn't have time to finish this properly. Adding the URL and/or handle (every asset has to have a unique handle -> `id='my-script'`) in the replacement, although not bullet proof, would harden things enough IMHO.
> 
> You check it here : [master...nlemoine:fix-output-filters](https://github.com/inpsyde/assets/compare/master...nlemoine:fix-output-filters)

`DOMDocument` can be solid approach, only downside can be performance here, if we can setup a test code to verify performance for `str_replace` vs `DOMDocument` large no. of scripts / styles.

**Comment by lkraav** [comment]
I checked https://github.com/inpsyde/assets/compare/master...nlemoine:fix-output-filters

Couldn't determine whether it makes sense to bring in `current_theme_supports()` complexity directly in here.

Pushed an update to #20 to `preg_match` handle any existing `type` attribute value, with an extra test to boot.

To me eye, seems like the currently clearest and simplest available approach.

DOMDocument measurement would be interesting to run, but I'd rather work on 3rd party asset control issue. I think with DOMDocument, we'd be talking some ms difference vs str_replace, which just doesn't justify much time spend outside scientific interest.

**Comment by Chrico** [comment]
Thanks a lot for the really good activity and effort put into this issue/pull! :) 

Sry to say that, but maybe i need you to ask to slow down a bit. Latest WordPress in version 5.7 introduced a new API to print scripts and inline scripts with various attributes [1]. 
Sadly this API is not yet used internally for `WP_Scripts::do_item()`, but i can foresee, that this will be used in future, to internally print scripts. I would like to wait a bit here and maybe find some time to check https://core.trac.wordpress.org/ [2] [3] to get a good overview about the directon they are planning to go.

This would maybe allow us to just implement a new public API via `Script::withAttributes()` and `Script::attributes()`, which will be used in `ScriptHandler` via those new filters. This way we could also get rid of the `DeferScriptOutputFilter` and `AsyncScriptOutputFilter` and solve the problem with adding various attributes. :) 

[1] https://make.wordpress.org/core/2021/02/23/introducing-script-attributes-related-functions-in-wordpress-5-7/
[2] https://core.trac.wordpress.org/ticket/12009
[3] https://core.trac.wordpress.org/

**Comment by lkraav** [comment]
> This would maybe allow us to just implement a new public API via Script::withAttributes() and Script::attributes(), which will be used in ScriptHandler via those new filters. 

Absolutely, but we could also have this as a separate project, because it's a large refactor I'm not sure I can make enough time for. If Inpsyde also doesn't have energy, this could delay a useful thing for months.

At the same time, I'd like to start using this.

If this PR gets stalled here, I'm aware I could also just implement this class in my own theme lib - little bit of misdirection, but certainly better than nothing. Then we'll see however long a larger 5.7-compatible refactor would take.

> Latest WordPress in version 5.7

On this topic, would it mean a new major Assets version, and break back-compat? Because you'd have to maintain current OutputFilters indefinitely, otherwise.

**Comment by nlemoine** [comment]
> Couldn't determine whether it makes sense to bring in current_theme_supports() complexity directly in here.

Tried to stick to `str_replace`. The issue with your implementation (which is basically the same as existing `async`/`defer` output filters except that it takes care of the `type="text/javascript"` attribute) is that you will end up with this:
```html
<script type="module" src='foo.js' id='handle'></script>
<script type="module" defer id="handle">
var foo = "bar"
</script>
```

in case you append some inline script (see https://github.com/inpsyde/assets/issues/6#issue-687267961)

**Comment by Chrico** [comment]
> bsolutely, but we could also have this as a separate project, because it's a large refactor I'm not sure I can make enough time for.

Actually its not a refactor, we just extend the public API. 
Basically we need to add to `BaseAsset` those 2 methods and internally collect "additional attributes" (aside from those which are avaiable via main API like src/href, media, ..). And in `ScriptHandler` and `StyleHandler` we need to handle those like localized data. Maybe there could be even a `AttributesOutputFilter` which is automatically added when the "additional attributes` are not empty.

But thats something which i just need a bit time to avoid rushing into a direction which lateron will provide BC.

> At the same time, I'd like to start using this.

Really understand your idea. Still you can have a small library with some helpers or even use it per project. 

> On this topic, would it mean a new major Assets version, and break back-compat?

No, i don't think so. Currently OutputFilter are applied to `script_loader_tag` and `style_loader_tag`. I don't think that WordPress will remove those in future. They will probably just implement into `WP_Scripts::do_item()` here https://github.com/WordPress/WordPress/blob/032aec8df2bf2ebdfd46dd19c6df5dd470a9ee4b/wp-includes/class.wp-scripts.php#L388 instead of building the `$tag` by hand.
The problem i see here is: How do they manage to add attributes to inline scripts which are printed before/after. Will they have the same attributes like the loaded one via `src` ? This might can break things if we have `async` on all script tags for a handle.

**Comment by lkraav** [comment]
> in case you append some inline script

There's no failing test case currently filed anywhere in some private branch, right?

But regardless, I'm currently live with such setup, for 12+ months, zero ill effects. I'm not sure having `<script type="module">` for inline hurts, and use cases for specifically avoiding it when main script is `type=module`, might be even less % than having any `type=module` scripts.

**Comment by nlemoine** [comment]
> There's no failing test case currently filed anywhere in some private branch, right?

Because this library only provides unit tests but no integration tests (see #16, ping @Chrico ;)).

Here's the [failing test](https://github.com/nlemoine/assets/runs/1864274344?check_suite_focus=true).

> I'm not sure having <script type="module"> for inline hurts

It probably doesn't but I prefer not relying on a "probably".

**Comment by Chrico** [comment]
> Because this library only provides unit tests but no integration tests (see #16, ping @Chrico ;)).

yep i know! And thats the whole case, currently we're really limited, because those 2 filters (for Style and Script) are throwing in everything, also inline script/styles before and after. So we have no way to _just_ filter the Script/Style which actually loads the external resource and should be a module, deferred or async. It's a limitation from WordPress which we cannot simply go arround, eventho it's an edge case right now. 

As i said: i need a bit time to investigate, also how they plan to implement the whole "CSP"-topic to core scripts which are having inline scripts before/after. :)

**Comment by lkraav** [comment]
> It probably doesn't but I prefer not relying on a "probably".

https://gist.github.com/jakub-g/385ee6b41085303a53ad92c7c8afd7a6 lowers "probably" to something like "near non-existent".

But, I can see I'll probably implement this in my own lib to get going, and look to chip in when 5.7 compatible upgrade work ever begins.

**Comment by Chrico** [comment]
Just a quick throw-in before i go off: I already started implementation of the concept i mentioned early here: https://github.com/inpsyde/assets/tree/asset-attributes :-)

**Comment by lkraav** [comment]
>  I already started implementation of the concept i mentioned early here: https://github.com/inpsyde/assets/tree/asset-attributes

Looking through it now. What do you feel is still missing from merge-ready?

**Comment by Chrico** [comment]
> What do you feel is still missing from merge-ready

Probably not much. Already tested a bit in various projects and looks good. Pushed right now updated documentation (README and DocBlocks) and also refactored a bit the internals of `Inpsyde\Assets\BaseAsset` to not rely anymore on a big `$config`-array.

Will give a shout in company to test this and might merge it this week if nothing "bad" occures :)

**Comment by lkraav** [comment]
> > What do you feel is still missing from merge-ready
> 
> Probably not much. Already tested a bit in various projects and looks good.

OK I'll also base my stuff on this branch for some external alpha testing and see if it works for us.

**Comment by Chrico** [comment]
Merged the branch into `master`. Will try to finalize documentation today/tomorrow and createa a release. Tests went good so far.

**Comment by Chrico** [comment]
Just released today `inpsde/assets` 2.5. See more here: https://github.com/inpsyde/assets/releases/tag/2.5 :-) 

Thanks all for your help!
